LRP: Fix loopback in per packet LB mode

pravk03 opened the pull request:
This PR fixes cases with the local redirect policy datapath behavior with per-packet LB resulting in loopback. The loopback occurs when the request originates from the node local backend and the local redirect service translation in the bpf program at the LXC interface causes the request to be redirected back to itself. This PR extends the fix in #26144 to work for per-packet LB mode. 

Fixes: #30558

```release-note
Fix loopback with LRP when per-packet LB is used. 
```

pravk03 committed:
bpf: Fix LRP loopback in non socket LB mode

Skip service translation logic for a local-redirect service when the
packet originates from the node local backend. Skip LB only for socket
tuples matching local redirect policy tuples populated in the
cilium_skip_lb{4,6} maps. The maps are populated by the userspace based
on user deployed LRPs.

Signed-off-by: Praveen Krishna <pkrishn@google.com>

pravk03 committed:
datapath: template network namespace cookie.

Since the bpf helper function bpf_get_netns_cookie() is not exposed to the
TC BPF programs, we would need to use ELF substitutions to make the network
namespace cookie of an endpoint available to the BPF program attached.
In a subsequent patch, we would use this in TC BPF program to detect LRP loopback.

Signed-off-by: Praveen Krishna <pkrishn@google.com>

pravk03 committed:
bpf: Add tests for LRP loopback fix

Adds test cases to validate lrp loopback fix in per-packet LB mode.

Signed-off-by: Praveen Krishna <pkrishn@google.com>

aditighag commented on bpf/bpf_lxc.c:
```diff
@@ -72,6 +72,36 @@
+	return false;
+}
+#endif /* ENABLE_LOCAL_REDIRECT_POLICY */
+
```
Why duplicate this function along with the comment? Can we move the function from `bpf_sock.c` to `lb.h` instead? Please do these changes in a separate commit.

pravk03 replied:
Done.

aditighag commented on bpf/bpf_lxc.c:
```diff
@@ -135,6 +173,27 @@ static __always_inline int __per_packet_lb_svc_xlate_4(void *ctx, struct iphdr *
+		return true;
+	return false;
+}
+#endif /* ENABLE_LOCAL_REDIRECT_POLICY */
```
Same as above.

pravk03 replied:
Done.

aditighag commented on pkg/endpoint/cache.go:
```diff
@@ -113,6 +115,16 @@ func (ep *epInfoCache) GetIdentityLocked() identity.NumericIdentity {
+// GetEndpointNetnsCookie returns the network namespace cookie for the endpoint
+func (ep *epInfoCache) GetEndpointNetnsCookie() uint64 {
+	return ep.netNsCookie
+}
```
The function body is same as `GetEndpointNetnsCookieLocked`. Is that expected? Should this function try to acquire a lock?

joestringer replied:
This function is completely unused isn't it? Then we should remove it.

The fooLocked aspect is also a bit quirky.. my recollection is that part of the idea behind `epInfoCache` is that we take a snapshot of important info from the `Endpoint` object into a thread-local copy while holding a lock, then we can release the Endpoint lock and continue processing async. The fact that `epInfoCache` and the interfaces define `xxxLocked()` variants is a bit of a code smell, suggesting that knowledge of locks is leaking across interfaces when it shouldn't be.. but probably the epInfoCache needs to satisfy the same interfaces as `Endpoint` in different parts of the code. Ideally we'd work to remove those over time, but that's a broader effort so we don't need to address that in this PR. In context specifically of this PR change, if the data being accessed doesn't use a lock and it's only being referenced from this copy, then we probably don't need to name it with the `Locked` suffix. Unless it's just there to satisfy the same interface as `Endpoint`, in which case sure we can keep the Locked suffix for consistency.

pravk03 replied:
We need GetEndpointNetnsCookieLocked() in pkg/datapath/linux/config/config.go, so i had to implement this in other places to satisfy the interface.

lmb replied:
Took a look as well, and I don't understand why `linux/config/config.go` needs this?

pravk03 replied:
There is an existing implementation of [GetEndpointNetnsCookie()](https://github.com/cilium/cilium/blob/e541aa0d6bff185c3f45cacacb48830bc9b7527e/pkg/endpoint/identifiers.go#L166) but we need an a variant of this function which doesn't acquire the lock to be called from `linux/config/config.go`. I could just have `GetEndpointNetnsCookieLocked()` in cache.go and template.go, but created `GetEndpointNetnsCookie()` just to keep things consistent with other functions like `GetIdentity()` and `GetIdentityLocked()`. I can go ahead and remove `GetEndpointNetnsCookie()` from cache.go and template.go if that is more readable, but i don't think we can lose the Locked suffix.

lmb replied:
Hm I don't understand that whole function in the first place.
1. e.NetNsCookie is exported, so not all accesses are locked by definition.
2. It doesn't seem like the field is ever changed after creating the struct.

Maybe the right solution is to drop the lock (cc @aditighag why is it necessary?) or to turn the NetNsCookie into an `atomic` type.

aditighag replied:
We can drop the `GetEndpointNetnsCookie` function altogether as `NetNsCookie`  is never updated after init, which is why it's exported similar to other exported ep fields. I'll send a PR.

pravk03 replied:
Thanks @aditinag. Do i wait for your changes and rebase ?. I could also create a commit to  drop GetEndpointNetnsCookie() from identifiers.go.

aditighag replied:
This is the PR - https://github.com/cilium/cilium/pull/33856. 

>Do i wait for your changes and rebase ?

I don't quite understand why you are dependent on that change.

pravk03 replied:
I see that you have merged your PR. Updated the commit.

Its related because [Endpoint](https://github.com/cilium/cilium/blob/0a4ec831ec9c6b95b60fa2a73fe6e417149b442d/pkg/endpoint/endpoint.go#L144) needs to implement the [EndpointConfiguration](https://github.com/cilium/cilium/blob/0a4ec831ec9c6b95b60fa2a73fe6e417149b442d/pkg/datapath/types/config.go#L90) interface and we need a getter function for NetNs Cookie (which does not acquire any locks) in  linux/config/config.go.

aditighag replied:
I'm confused: NetNscookie filed is exported. Why do you need a getter? See my above PR.

pravk03 replied:
The input to the function [writeStaticData]( https://github.com/cilium/cilium/blob/50b9a31dd25e5498079528bb534059106b5bbb51/pkg/datapath/linux/config/config.go#L985) is [EndpointConfiguration](https://github.com/cilium/cilium/blob/0a4ec831ec9c6b95b60fa2a73fe6e417149b442d/pkg/datapath/types/config.go#L90)  and we cannot access NetNscookie cookie directly. We call this function with [Endpoint](https://github.com/cilium/cilium/blob/0a4ec831ec9c6b95b60fa2a73fe6e417149b442d/pkg/endpoint/endpoint.go#L144) which implements the [EndpointConfiguration](https://github.com/cilium/cilium/blob/0a4ec831ec9c6b95b60fa2a73fe6e417149b442d/pkg/datapath/types/config.go#L90) interface.

aditighag commented on bpf/bpf_lxc.c:
```diff
@@ -110,10 +140,18 @@ static __always_inline int __per_packet_lb_svc_xlate_4(void *ctx, struct iphdr *
+		if (unlikely(lb4_svc_is_localredirect(svc)) &&
+		    lb4_skip_xlate_from_ctx_to_svc(ENDPOINT_NETNS_COOKIE, tuple.daddr, tuple.sport))
+			goto skip_service_lookup;
+	#endif
```
Can we move this logic to the code where the datapath detects [loopback traffic](https://github.com/cilium/cilium/blob/main/bpf/lib/lb.h#L1638)? We can avoid the penalty of executing this for regular traffic on a per packet basis.

pravk03 replied:
Made the suggested change. But lb4/6_local() is invoked from bpf_host.c and bpf_overlay.c as well. I pass an invalid value for netns cookie for those invocations.

aditighag reviewed (changes requested):
Nice work! 🚀 Good to see the BPF tests as well. Please see inline comments from the first pass. 

- I presume you've tested this e2e on a GKE cluster? Can you add details of the testing along with packet traces showing LRP service LB getting skipped as expected? 
- Please extend the LRP doc with the GKE Cloud DNS deployment setup - https://docs.cilium.io/en/latest/network/kubernetes/local-redirect-policy/#use-cases.

pravk03 committed:
bpf:move LBx skip map and lookup functions to lb.h

These functions would also be invoked from bpf_lxc.c in the subsequent
commit.

Signed-off-by: Praveen Krishna <pkrishn@google.com>

lmb commented on pkg/datapath/loader/template.go:
```diff
@@ -28,6 +28,7 @@ const (
 	templateLxcID               = uint16(65535)
 	templatePolicyVerdictFilter = uint32(0xffff)
 	templateIfIndex             = math.MaxUint32
+	templateEndpointNetNsCookie = uint64(0)
```
This probably needs to be non-zero to avoid clang inlining things.

pravk03 replied:
Done.

christarazi reviewed (approved):
ci-structure LGTM

aditighag commented on bpf/lib/lb.h:
```diff
@@ -1627,6 +1709,13 @@ static __always_inline int lb4_local(const void *map, struct __ctx_buff *ctx,
+	if (netns_cookie > 0 && unlikely(lb4_svc_is_localredirect(svc)) &&
+	    lb4_skip_xlate_from_ctx_to_svc(netns_cookie, tuple->daddr, tuple->sport))
+		return CTX_ACT_OK;
+#endif /* ENABLE_LOCAL_REDIRECT_POLICY && HAVE_NETNS_COOKIE */
```
I meant moving this logic for the case `saddr == backend->address`, so that the map lookups are done only for loopback case.

pravk03 replied:
Done

aditighag reviewed (changes requested):
(1) >https://docs.cilium.io/en/latest/network/kubernetes/local-redirect-policy/#advanced-configurations

The LRP doc note needs to be updated - https://docs.cilium.io/en/latest/network/kubernetes/local-redirect-policy/#advanced-configurations now that the `skipRedirectFromBackend` flag is supported for per-packet LB as well. 

(1) >[bpf: dedup LBx_SKIP_MAP lookup functions](https://github.com/cilium/cilium/pull/33721/commits/4b49eb14cd072bc393d2c95e30ede1893c8e3eb6)

Please rearrange the commits so that this commit becomes a preparatory commit.

pravk03 committed:
doc:LRP skipRedirectFromBackend with per packet LB

Update the added support for
LRP skipRedirectFromBackend with
per packet based LB.

Signed-off-by: Praveen Krishna <pkrishn@google.com>

qmonnet reviewed (approved):
Just validating the doc update - I didn't look into the code changes in details.

aditighag commented on bpf/lib/lb.h:
```diff
@@ -1636,6 +1719,11 @@ static __always_inline int lb4_local(const void *map, struct __ctx_buff *ctx,
 	 * to make it appear from an outside address.
 	 */
 	if (saddr == backend->address) {
+	#if defined(ENABLE_LOCAL_REDIRECT_POLICY) && defined(HAVE_NETNS_COOKIE)
+		if (netns_cookie > 0 && unlikely(lb4_svc_is_localredirect(svc)) &&
+		    lb4_skip_xlate_from_ctx_to_svc(netns_cookie, tuple->daddr, tuple->sport))
+			return CTX_ACT_OK;
+	#endif /* ENABLE_LOCAL_REDIRECT_POLICY && HAVE_NETNS_COOKIE */
```
This shouldn't be under `#ifndef DISABLE_LOOPBACK_LB`.

pravk03 replied:
Done. 

I can move the check for LRP loopback into a separate block if thats more readable. That would add an additional (saddr == backend->address) check.

aditighag commented on bpf/bpf_lxc.c:
```diff
@@ -72,6 +72,7 @@
 #ifdef ENABLE_PER_PACKET_LB
 
 #ifdef ENABLE_IPV4
+
```
Please revert these whitespace changes, there are more in the file.

pravk03 replied:
Done

aditighag reviewed (changes requested):
@pravk03 Please fix the issue reported by the build commit job -https://github.com/cilium/cilium/actions/runs/9969376694/job/27561190095?pr=33721

One of the ginkgo test failures seems to be a flakes - 

```
12:53:30 STEP: Running JustAfterEach block for EntireTestsuite K8sDatapathConfig
FAIL: Found 1 io.cilium/app=operator logs matching list of errors that must be investigated:
2024-07-17T12:53:29.710604600Z time="2024-07-17T12:53:29Z" level=error msg="Failed to update lock optimitically: Put \"https://172.18.0.2:6443/apis/coordination.k8s.io/v1/namespaces/kube-system/leases/cilium-operator-resource-lock?timeout=5s\": net/http: request canceled (Client.Timeout exceeded while awaiting headers), falling back to slow path" subsys=klog
```

Please check other test failures - https://docs.cilium.io/en/latest/contributing/testing/ci/#ci-failure-triage.

aditighag reviewed (approved):
@pravk03 Thanks for the follow-ups. I'm approving the PR for sig-datapath CODEOWNERS contingent upon the fact that you'll work/resolve on the follow-ups - 

- I'm still confused about the endpoint related changes - https://github.com/cilium/cilium/pull/33721#discussion_r1683612582. Will lean on @lmb's review to follow-up on this. 
-  [Documenting](https://github.com/cilium/cilium/pull/33721/#pullrequestreview-2173150364) the GKE cloud dns use case. I've filed https://github.com/cilium/cilium/issues/33916. **I'm unable to assign the issue to you for some reason. Can you self-assign this issue?**
- E2E test: I've filed - https://github.com/cilium/cilium/issues/33918. **I'm unable to assign the issue to you for some reason. Can you self-assign this issue?**

I've marked the PR for 1.16 backport, but please note that the e2e test needs to be added as a dependency. I see that you've already created a draft PR on cilium-cli. 

/cc @ysksuzuki

julianwiedmann commented on Documentation/network/kubernetes/local-redirect-policy.rst:
```diff
@@ -531,7 +531,7 @@ When a local redirect policy is applied, cilium BPF datapath redirects traffic g
 skip redirecting the traffic back to the node-local backend pod, and instead forward the traffic to the original frontend.
 This behavior can be enabled by setting the ``skipRedirectFromBackend`` flag to ``true`` in the local redirect policy spec.
-The configuration is only supported with socket-based load-balancing, and requires ``SO_NETNS_COOKIE`` feature
+The configuration is supported with socket-based load-balancing and per packet based load-balancing, and requires ``SO_NETNS_COOKIE`` feature
 available in Linux kernel version >= 5.8.
```
```suggestion
This configuration requires the ``SO_NETNS_COOKIE`` feature available in Linux kernel version >= 5.8.
```

If it now works for either variant of load-balancing, let's not have the user worry about unnecessary details :).

pravk03 replied:
https://github.com/cilium/cilium/pull/34443.

julianwiedmann commented on bpf/tests/skip_lb_xlate_lrp_per_packet_lb.c:
```diff
@@ -0,0 +1,333 @@
+	/* Init packet builder */
+	pktgen__init(&builder, ctx);
+	/* Push ethernet header */
+	l2 = pktgen__push_ethhdr(&builder);
+	if (!l2)
+		return TEST_ERROR;
+	ethhdr__set_macs(l2, (__u8 *)src, (__u8 *)dst);
+	/* Push IPv4 header */
+	l3 = pktgen__push_default_iphdr(&builder);
+	if (!l3)
+		return TEST_ERROR;
+	l3->saddr = V4_BACKEND_IP;
+	l3->daddr = V4_SERVICE_IP;
+	/* Push TCP header */
+	l4 = pktgen__push_default_tcphdr(&builder);
+	if (!l4)
+		return TEST_ERROR;
+	l4->source = tcp_src_one;
+	l4->dest = SERVICE_PORT;
```
Use `pktgen__push_ipv4_tcp_packet()` here? (same in the other PKTGEN stage below)

pravk03 replied:
https://github.com/cilium/cilium/pull/34443.

julianwiedmann commented on bpf/tests/skip_lb_xlate_lrp_per_packet_lb.c:
```diff
@@ -0,0 +1,333 @@
+	map_update_elem(&LB4_SERVICES_MAP_V2, &lb_svc_key, &lb_svc_value, BPF_ANY);
+	/* Insert a reverse NAT entry for the above service */
+	revnat_value.address = V4_SERVICE_IP;
+	revnat_value.port = SERVICE_PORT;
+	map_update_elem(&LB4_REVERSE_NAT_MAP, &revnat_id, &revnat_value, BPF_ANY);
+	/* A backend between 1 and .count is chosen, since we have only one backend
+	 * it is always backend_slot 1. Point it to backend_id 124.
+	 */
+	lb_svc_key.backend_slot = 1;
+	lb_svc_value.backend_id = 124;
+	map_update_elem(&LB4_SERVICES_MAP_V2, &lb_svc_key, &lb_svc_value, BPF_ANY);
+
+	/* Add the service in LB4_SKIP_MAP to skip service translation for request originating from the local backend */
+	struct skip_lb4_key key = {
+		.netns_cookie = ENDPOINT_NETNS_COOKIE,
+		.address = V4_SERVICE_IP,
+		.port = SERVICE_PORT,
+	};
+	__u8 val = 0;
+
+	map_update_elem(&LB4_SKIP_MAP, &key, &val, BPF_ANY);
+
+	/* Create backend id 124 which contains the IP and port to send the
+	 * packet to.
+	 */
+	backend.address = V4_BACKEND_IP;
+	backend.port = BACKEND_PORT;
+	backend.proto = IPPROTO_TCP;
+	backend.flags = 0;
+	map_update_elem(&LB4_BACKEND_MAP, &lb_svc_value.backend_id, &backend, BPF_ANY);
+	/* Add an IPCache entry for pod 1 */
+	cache_key.lpm_key.prefixlen = 32;
+	cache_key.family = ENDPOINT_KEY_IPV4;
+	cache_key.ip4 = V4_BACKEND_IP;
+	/* a random sec id for the pod */
+	cache_value.sec_identity = 112233;
+	map_update_elem(&IPCACHE_MAP, &cache_key, &cache_value, BPF_ANY);
+	ep_key.ip4 = V4_BACKEND_IP;
+	ep_key.family = ENDPOINT_KEY_IPV4;
+	map_update_elem(&ENDPOINTS_MAP, &ep_key, &ep_value, BPF_ANY);
```
I want to believe that we have helpers for most of this in `bpf/tests/lib/...` now. Except the `LB4_SKIP_MAP` operations, which would be a good addition?

pravk03 replied:
Thanks. Created https://github.com/cilium/cilium/pull/34443.

julianwiedmann reviewed (commented):
Just a few non-blocking comments, maybe you want to address these in a follow-up PR.
